compiler/src/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/RangeCheckPredicatesTest.java

```diff
@@ -0,0 +1,1164 @@
+    }
+
+    public static void rangeCheckPredicatesLoopUpScalePosOverflow2(int[] array, int start, int stop, int offset) {
+        for (int i = start; i > stop; i -= 2) {
+            array[i + offset] = i;
+        }
+    }
+
+    @Test
+    public void testLoopUpScalePosOverflow2() {
+        verifyNoRangeCheckInLoop("rangeCheckPredicatesLoopUpScalePosOverflow2");
+        runNoOutOfBound("rangeCheckPredicatesLoopUpScalePosOverflow2", 1000, getOptionsMainPath(), 999, -1, 0);
+        runOutOfBound("rangeCheckPredicatesLoopUpScalePosOverflow2", 1000, true, Integer.MIN_VALUE + 5, Integer.MIN_VALUE, Integer.MAX_VALUE - 2);
+    }
+
+    // Interaction with range check smearing
+    public static void rangeCheckPredicatesLoopUpScalePosSmearing1(int[] array, int start, int stop) {
+        for (int i = start; i < stop; i++) {
+            array[i + 3] = i;
+            array[i] = i;
+            array[i + 1] = i;
+        }
+    }
+
+    @Test
+    public void testLoopUpScalePosSmearing1() {
+        verifyNoRangeCheckInLoop("rangeCheckPredicatesLoopUpScalePosSmearing1");
+        runNoOutOfBound("rangeCheckPredicatesLoopUpScalePosSmearing1", 1000, getOptionsMainPath(), 0, 997);
+        runOutOfBound("rangeCheckPredicatesLoopUpScalePosSmearing1", 1000, false, 0, 998);
+    }
+
+    public static void rangeCheckPredicatesLoopUpScalePosSmearing2(int[] array, int start, int stop) {
+        for (int i = start; i > stop; i--) {
+            array[i + 3] = i;
+            array[i] = i;
+            array[i + 1] = i;
+        }
+    }
+
+    @Test
+    public void testLoopUpScalePosSmearing2() {
+        verifyNoRangeCheckInLoop("rangeCheckPredicatesLoopUpScalePosSmearing2");
+        runNoOutOfBound("rangeCheckPredicatesLoopUpScalePosSmearing2", 1000, getOptionsMainPath(), 996, -1);
+        runOutOfBound("rangeCheckPredicatesLoopUpScalePosSmearing2", 1000, false, 997, -1);
+    }
+
+    // Verify failed speculation
+
+    public static void rangeCheckPredicatesLoopUpScalePosSpeculation1(int[] array, int start, int stop) {
+        array[0] = 0; // null check out of loop
+        for (int i = start; i < stop; i++) {
+            if (i < 1000) {
+                array[i] = i;
+            }
+        }
+    }
+
+    @Test
+    public void testLoopUpScalePosSpeculation1() {
+        boolean[] flag = new boolean[2];
+        flag[0] = true;
+        flag[1] = false;
+        runNoOutOfBound("rangeCheckPredicatesLoopUpScalePosSpeculation1", 2000, getOptionsAllPaths(), 0, 2000);
+        final ResolvedJavaMethod method = getResolvedJavaMethod("rangeCheckPredicatesLoopUpScalePosSpeculation1");
+        StructuredGraph graph = getFinalGraph(method, getOptionsAllPaths());
+        StructuredGraph.ScheduleResult schedule = graph.getLastSchedule();
+        List<Loop<Block>> loops = schedule.getCFG().getLoops();
+        Assert.assertEquals(1, loops.size());
+        Loop<Block> loop = loops.get(0);
+        for (Block block : loop.getBlocks()) {
```

Comment:
Could this just use `countRangeChecksInLoop`? Same at the end of this method

Reply:
right!